File: lona/html/canvas/canvas.js
function LonaCanvasWidget(lona_window) {
    this.lona_window = lona_window;

    this._run_operations = function() {
        for(var index in this.data['operations']) {
            var operation = this.data['operations'][index];

            // set
            if(operation[0] == 'set') {
                eval('this.ctx.' + operation[1] + ' = operation[2];');

            // call
            } else if(operation[0] == 'call') {
                eval('this.ctx.' + operation[1] + '(...operation[2])');

            };
        };
    };

    this.setup = function() {
        var lona_window = this.lona_window;

        this.canvas = this.nodes[0];
        this.ctx = this.canvas.getContext(this.data['context']);

        this._run_operations();
    };

    this.data_updated = function() {
        this._run_operations();
    };
};

Lona.register_widget_class('LonaCanvasWidget', LonaCanvasWidget);

File: lona/html/canvas/canvas.py
from lona.static_files import Script
from lona.html.widget import Widget
from lona.html.node import Node


class CanvasNode(Node):
    TAG_NAME = 'canvas'


class CanvasContext:
    def __init__(self, widget):
        self._widget = widget

    def _gen_callback(self, name):
        def callback(*args):
            self._widget.data['operations'].append(
                ['call', name, list(args)],
            )

        return callback

    def __getattribute__(self, name):
        if name.startswith('_'):
            return super().__getattribute__(name)

        return self._gen_callback(name)

    def __setattr__(self, name, value):
        if name.startswith('_'):
            return super().__setattr__(name, value)

        self._widget.data['operations'].append(
            ['set', name, value],
        )


class Canvas(Widget):
    FRONTEND_WIDGET_CLASS = 'LonaCanvasWidget'

    STATIC_FILES = [
        Script(name='LonaCanvasWidget', path='canvas.js'),
    ]

    def __init__(self, *args, **kwargs):
        self.nodes = [
            CanvasNode(*args, **kwargs),
        ]

        self.data = {
            'context': '2d',
            'operations': [],
        }

        self.ctx = CanvasContext(widget=self)
